lib/src/gitattributes.rs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_docs)]
+
+use gix::attrs as gix_attrs;
+use gix::glob as gix_glob;
+use gix::path as gix_path;
+use std::borrow::Cow;
+use std::io;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum GitAttributesError {
+    #[error("Failed to read attributes patterns from file {path}")]
+    ReadFile { path: PathBuf, source: io::Error },
+}
+
+/// Models the effective contents of multiple .gitattributes files.
+#[derive(Debug)]
+pub struct GitAttributesFile {
+    search: gix_attrs::Search,
+    collection: gix_attrs::search::MetadataCollection,
+}
+
+impl GitAttributesFile {
+    pub fn empty() -> Arc<GitAttributesFile> {
+        Arc::new(GitAttributesFile {
+            search: gix_attrs::Search::default(),
+            collection: gix_attrs::search::MetadataCollection::default(),
+        })
+    }
+
+    pub fn chain(
+        self: &Arc<GitAttributesFile>,
+        prefix: PathBuf,
+        input: &[u8],
+    ) -> Result<Arc<GitAttributesFile>, GitAttributesError> {
+        let mut source_file = prefix.clone();
+        source_file.push(".gitattributes");
+
+        let mut search = self.search.clone();
+        let mut collection = self.collection.clone();
+        search.add_patterns_buffer(input, source_file, Some(&prefix), &mut collection, true);
+        Ok(Arc::new(GitAttributesFile { search, collection }))
+    }
```

Comment:
I wound up doing something similar myself when looking at testing this (https://github.com/kejadlen/jj/commit/5c423f84a5be1e268fb43c33772c3356fab1ac54) but I'm not particularly fond of not testing the path that's actually being exercised in actual usage, so have been mulling on a satisfactory alternative.

For manual testing, I've been using [this](https://github.com/kejadlen/jj-lfs-test/commit/b863adeaec5d549f2fef5dd1af5a0baa78844193) and was wondering if it's worthwhile to package this up and write a more integration-y level test? 

```sh
❯ touch a.png b.jpg foo/c.png foo/d.jpg
❯ jj st
Working copy changes:
A b.jpg
Working copy : wuvuwzpo c3e9dd78 (no description set)
Parent commit: oznmowzq b863adea μ
```

Reply:
Although that being said, I think I've found a bug in that I've only tested with new files so far, and we probably need to exclude also already tracked files when snapshotting?

Reply:
Alright, splitting out the LFS conditional in `process_dir_entry` seems to do the trick, although it is a somewhat confusing user experience, with `jj st` showing that the LFS files will be deleted (although not actually doing so in the commit).